kornia/geometry/epipolar/essential.py

```diff
@@ -19,6 +22,181 @@
+# https://github.com/weitong8591/differentiable_ransac/blob/main/estimators/essential_matrix_estimator_nister.py
+
+
+def run_5point(points1: torch.Tensor, points2: torch.Tensor, weights: Optional[torch.Tensor] = None) -> torch.Tensor:
+    r"""Compute the essential matrix using the 5-point algorithm from Nister.
+
+    The linear system is solved by Nister's 5-point algorithm.
+
+    Args:
+        points1: A set of carlibrated points in the first image with a tensor shape :math:`(B, N, 2), N>=8`.
+        points2: A set of points in the second image with a tensor shape :math:`(B, N, 2), N>=8`.
+        weights: Tensor containing the weights per point correspondence with a shape of :math:`(B, N)`.
+
+    Returns:
+        the computed fundamental matrix with shape :math:`(B, 3, 3)`.
+    """
+    KORNIA_CHECK_SHAPE(points1, ['B', 'N', '2'])
+    KORNIA_CHECK_SAME_SHAPE(points1, points2)
+    KORNIA_CHECK(points1.shape[1] >= 5, "Number of points should be >=5")
+    if weights is not None:
+        KORNIA_CHECK_SAME_SHAPE(points1[:, :, 0], weights)
+
+    batch_size, _, _ = points1.shape
+    x1, y1 = torch.chunk(points1, dim=-1, chunks=2)  # Bx1xN
+    x2, y2 = torch.chunk(points2, dim=-1, chunks=2)  # Bx1xN
+    ones = torch.ones_like(x1)
+
+    # build equations system and solve DLT
+    # https://www.cc.gatech.edu/~afb/classes/CS4495-Fall2013/slides/CS4495-09-TwoViews-2.pdf
+    # [x * x', x * y', x, y * x', y * y', y, x', y', 1]
+    # BxNx9
+    X = torch.cat([x1 * x2, x1 * y2, x1, y1 * x2, y1 * y2, y1, x2, y2, ones], dim=-1)
+
+    # apply the weights to the linear system
+    if weights is None:
+        X = X.transpose(-2, -1) @ X
+    else:
+        w_diag = torch.diag_embed(weights)
+        X = X.transpose(-2, -1) @ w_diag @ X
+    # compute eigevectors and retrieve the one with the smallest eigenvalue, using SVD
+    # turn off the grad check due to the unstable gradients from SVD.
+    # several close to zero values of eigenvalues.
+    _, _, V = _torch_svd_cast(X)  # torch.svd
+    null_ = V[:, :, -4:]  # the last four rows
+    nullSpace = V.transpose(-1, -2)[:, -4:, :]
+
+    coeffs = torch.zeros(batch_size, 10, 20, device=null_.device, dtype=null_.dtype)
+    d = torch.zeros(batch_size, 60, device=null_.device, dtype=null_.dtype)
+
+    # fun = lambda i, j : null_[:, 3 * j + i]
+    def fun(i: int, j: int) -> torch.Tensor:
+        return null_[:, 3 * j + i]
+
+    # Determinant constraint
+    coeffs[:, 9] = (
+        epi.numeric.o2(epi.numeric.o1(fun(0, 1), fun(1, 2)) - epi.numeric.o1(fun(0, 2), fun(1, 1)), fun(2, 0))
+        + epi.numeric.o2(epi.numeric.o1(fun(0, 2), fun(1, 0)) - epi.numeric.o1(fun(0, 0), fun(1, 2)), fun(2, 1))
+        + epi.numeric.o2(epi.numeric.o1(fun(0, 0), fun(1, 1)) - epi.numeric.o1(fun(0, 1), fun(1, 0)), fun(2, 2))
+    )
+
+    indices = torch.tensor([[0, 10, 20], [10, 40, 30], [20, 30, 50]])
+
+    # Compute EE^T (Eqn. 20 in the paper)
+    for i in range(3):
```

Comment:
wondering whether we could vectorise all the for loops below via gather or similar

Reply:
Let's do in follow-up PR